### PR TITLE
#312 restx-samplest test fix

### LINKED ISSUE #312: restx-samplest tests are failing
*Jdk version:* `java-1.8.0-openjdk-1.8.0.242.b08-0.fc31.x86_64`
*Maven version:* Maven plugin of Intellij Idea 2020.1 located in `<idea_path>/plugins/maven/lib/maven3` -> Maven 3.6.1 (core)

**Steps to reproduce:**
1. Checkout master branch (current commit: c92e983)
2. Apply workarond in #310
3. Perform a `mvn verify` on `restx-parent`

**Expected behaviour:**
Dependencies are downloaded, application is packaged, tests are launched

**Current behaviour:**
Error during `restx-samplest` tests execution:

```
Failed tests:
  AutoStartableTest.should_handle_auto_startable_in_dev_mode:31 expected:<...called: 1 started: 1[ closed: 0 instanciated: ]1 serverId: Jetty8#1...> but was:<...called: 1 started: 1[11 closed: 110 instanciated: 11]1 serverId: Jetty8#1...>
Tests in error:
  ParametersResourceSpecTest.samplest.core.ParametersResourceSpecTest » MongoTimeout
  UUIDResourceSpecTest.samplest.core.UUIDResourceSpecTest » MongoTimeout Timed o...
  SecurityResourceSpecsTest.samplest.security.SecurityResourceSpecsTest » MongoTimeout
  ValidationResourceTest.samplest.validation.ValidationResourceTest » MongoTimeout
  MongoResourceSpecsTest.objectsWithNewJongoAnnotations persistence » MongoTimeout
  MongoResourceSpecsTest.objectsWithObjectId persistence » MongoTimeout Timed ou...
  MongoResourceSpecsTest.objectsWithIdAnnotation persistence » MongoTimeout Time...
  MongoResourceSpecsTest.objectsWithObjectIdAnnotation persistence » MongoTimeout
  MongoResourceSpecsTest.samplest.jongo.MongoResourceSpecsTest » MongoTimeout Ti...
Tests run: 87, Failures: 1, Errors: 9, Skipped: 0
```

### PR DESCRIPTION
**Pull request by nhumblot**
Fix #312 

These changes fix an `mvn verify` failure on `restx-parent`. Moving some tests under `MongoRestxSpecTestsRunner.class` is not the most efficient thing but it has the advantage to make tests pass and allow contributors to focus on other issues. Should an issue be created regarding this `RestxSpecTestsRunner` & `MongoRestxSpecTestsRunner` thing ?

### CODE REVIEW

File: restx-specs-tests/src/main/java/restx/tests/RestxSpecTestsRunner.java [hunk 1]
```diff
@@ -48,7 +50,10 @@ public RestxSpecTestsRunner(Class<?> testClass) throws InitializationError, Ille
         super(testClass);
         FindSpecsIn findSpecsIn = getTestClass().getJavaClass().getAnnotation(FindSpecsIn.class);
         if (findSpecsIn != null) {
-            tests = new RestxSpecTests(new RestxSpecRule(findSpecsIn.withRouterPath()), RestxSpecTests.findSpecsIn(findSpecsIn.value()));
+            Factory factory = Factory.newInstance();
+            WebServerSupplier supplier = factory.queryByClass(WebServerSupplier.class).findOne().get().getComponent();
+            RestxSpecRule rule = new RestxSpecRule(findSpecsIn.withRouterPath(), supplier, factory);
+            tests = new RestxSpecTests(rule, RestxSpecTests.findSpecsIn(findSpecsIn.value()));
```
[the remark this answers is not in the document]
Comment (nhumblot) [reply restx-specs-tests/src/main/java/restx/tests/RestxSpecTestsRunner.java:56]: This merge request is not for java 11. These tests are failing with java 8. I have not tested them with a jdk11 to be honest, I first wanted to make the tests pass on java 8.

I provide elements in [my second comment](https://github.com/restx/restx/issues/312#issuecomment-612996458) in #312:

>It appears that GivenJongoCollectionSpecRuleSupplier is instantiated only once when executing the Maven job, at the first execution of a test running with RestxSpecTestsRunner or MongoRestxSpecTestsRunner.
>
>In my case, the first one was ParametersResourceSpecTest, which is not using MongoRestxSpecTestsRunner. So without an embedded Mongo being launched, the value in the settings is set to it's default value: localhost:21017, and stay like this for the whole testing suite.
>
>Because embedded Mongo is started for every class, we must refresh these settings for every test class, I had to create a new Factory instance for every RestxSpecTests (see: nhumblot@922ec78)

[the remark this answers is not in the document]
Comment (xhanin) [reply restx-specs-tests/src/main/java/restx/tests/RestxSpecTestsRunner.java:56]: Thanks, I missed this comment.
So I'm ok with this change on the runner.

File: restx-samplest/src/main/java/samplest/autostartable/AutoStartableTestComponent.java [hunk 2]
```diff
@@ -14,9 +14,9 @@
 */
 @Component
 public class AutoStartableTestComponent implements AutoStartable, AutoCloseable {
-    private static int closed;
-    private static int started;
-    private static int instanciated;
+    private int closed;
+    private int started;
+    private int instanciated;
```
[the remark this answers is not in the document]
Comment (xhanin) [reply restx-samplest/src/main/java/samplest/autostartable/AutoStartableTestComponent.java:19]: I still think we shouldn't change that. We should better change the way the test is called in `AutoStartableTest`, maybe using a new factory there would be enough.

**Comment by xhanin** [comment]
About the `MongoRestxSpecTestsRunner`, I still think we should find a better way. I think Mongo should be started only once for all tests - and the rule would just pick a up a new db for each test.

I think this implies moving to junit 5 which has @beforeall and @afterall and reviewing the logic (mongo embed rule should start and stop a db only if one is not already started globally by a beforeall). That's more work, but it's much cleaner.

WDYT?

File: restx-specs-tests/src/main/java/restx/tests/RestxSpecTestsRunner.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nhumblot) [reply restx-specs-tests/src/main/java/restx/tests/RestxSpecTestsRunner.java:56]: Ok, thank you. I will reply to the other two comments also and try to give as many details as I can. I will come back to you when I get the time to write them. :)

**Comment by nhumblot** [comment]
> Do you have any idea why moving to `MongoRestxSpecTestsRunner` makes test pass? It sounds like a ugly hack, it would imply accepting that `RestxSpecTestsRunner` is not working anymore. But I don't see that the mongo version brings to the table, that's weird.

I definitely see my changes as a workaround and not a long term solution. My commits message are not appropriates, sorry. As a user, I expected `RestxSpecTestsRunner` to work when testing a Controller not relying on a database connection. Starting a database while there is no need for it generates a big hit on test performance.

The fix seems to be long and hard to come with and I first focused on finding a workaround in order to have tests passing with jdk8 in order to test for jdk11 compatibility. This can allow someone to pursue a fix for this while having someone else setting up a CI and/or working on other enhancements/fixes. **If these changes are ever merged, it still does not solve the fact that `RestxSpecTestsRunner` is broken if `restx-jongo-specs-tests` is a dependency and I think this should be addressed anyway (now or in another pull request) because the behavior is not what to expect in my opinion.**

# TLDR

As soon as `restx-jongo-specs-tests` is imported as a dependency, `GivenJongoCollectionSpecRuleSupplierFactoryMachine` is in the classpath so `Factory` uses it to instantiate `GivenJongoCollectionSpecRuleSupplier` which instantiate `GivenJongoCollectionSpecRule` which inevitably need a Mongo instance due to `onTearDown()`. There is another issue at startup because the server try to establish a connection with a MongoDB instance while none as been started with `RestxSpecTestsRunner`.

Intellij Idea seems to ignore it if you execute tests in it. You still get a hit due to the connection at startup not happening but `onTearDown()` does not make the test fail. Maven does not like that and the test is failing because of the `onTearDown()` not succeeding.

**If `restx-jongo-specs-tests` is added as a dependency, `@RunWith(RestxSpecTestsRunner.class)` inevitably fail.**

**If you want me to, I can try to find the first commit that introduced the issue.**

# Test failure

When running `mvn verify`, I'm getting these errors:
```
Results :
Failed tests: 
  AutoStartableTest.should_handle_auto_startable_in_dev_mode:31 expected:<...called: 1 started: 1[ closed: 0 instanciated: ]1 serverId: Jetty8#1...> but was:<...called: 1 started: 1[11 closed: 110 instanciated: 11]1 serverId: Jetty8#1...>
Tests in error: 
  ParametersResourceSpecTest.samplest.core.ParametersResourceSpecTest » MongoTimeout
  UUIDResourceSpecTest.samplest.core.UUIDResourceSpecTest » MongoTimeout Timed o...
  SecurityResourceSpecsTest.samplest.security.SecurityResourceSpecsTest » MongoTimeout
  ValidationResourceTest.samplest.validation.ValidationResourceTest » MongoTimeout
  MongoResourceSpecsTest.objectsWithNewJongoAnnotations persistence » MongoTimeout
  MongoResourceSpecsTest.objectsWithObjectId persistence » MongoTimeout Timed ou...
  MongoResourceSpecsTest.objectsWithMongoIdAnnotation persistence » MongoTimeout
  MongoResourceSpecsTest.objectsWithIdAnnotation persistence » MongoTimeout Time...
  MongoResourceSpecsTest.objectsWithObjectIdAnnotation persistence » MongoTimeout
  MongoResourceSpecsTest.samplest.jongo.MongoResourceSpecsTest » MongoTimeout Ti...
Tests run: 88, Failures: 1, Errors: 10, Skipped: 0
```

# Stack trace

When looking at the execution of `UUIDResourceSpecTest`, we have the following logs:
```
Running samplest.core.UUIDResourceSpecTest
starting server
2020-04-28 21:15:38,121 [main            ] [          ] INFO  org.eclipse.jetty.server.Server - jetty-8.1.8.v20121106
LoginService=HashLoginService[null] identityService=org.eclipse.jetty.security.DefaultIdentityService@325bb9a6
2020-04-28 21:15:38,156 [main            ] [          ] INFO  restx.RestxMainRouterFactory - LOADING MAIN ROUTER
2020-04-28 21:15:38,156 [main            ] [          ] INFO  restx.RestxMainRouterFactory - 
--------------------------------------
 -- RESTX >> LOAD ON REQUEST << >> CLEAN << >> TEST MODE <<
 -- for admin console,
 --   VISIT http://127.0.0.1:33825/api/@/ui/
 --
server started
using db mongodb://localhost:27017/testing-mongo-test-1588108538174-2
---------------------------------------------------------------------------------
>> REQUEST
GET http://127.0.0.1:33825/api/uuids/random
2020-04-28 21:15:38,232 [qtp1225970981-189] [          ] INFO  org.mongodb.driver.cluster - Cluster created with settings {hosts=[localhost:27017], mode=SINGLE, requiredClusterType=UNKNOWN, serverSelectionTimeout='30000 ms', maxWaitQueueSize=500}
2020-04-28 21:15:38,237 [cluster-ClusterId{value='5ea89cfa533eb82c6c478c8b', description='null'}-localhost:27017] [          ] INFO  org.mongodb.driver.cluster - Exception in monitor thread while connecting to server localhost:27017
com.mongodb.MongoSocketOpenException: Exception opening socket
	at com.mongodb.internal.connection.SocketStream.open(SocketStream.java:70)
	at com.mongodb.internal.connection.InternalStreamConnection.open(InternalStreamConnection.java:128)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.run(DefaultServerMonitor.java:117)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connexion refusée (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:607)
	at com.mongodb.internal.connection.SocketStreamHelper.initialize(SocketStreamHelper.java:64)
	at com.mongodb.internal.connection.SocketStream.initializeSocket(SocketStream.java:79)
	at com.mongodb.internal.connection.SocketStream.open(SocketStream.java:65)
	... 3 common frames omitted
2020-04-28 21:15:38,238 [qtp1225970981-189] [          ] INFO  s.lifecycle.AutoCloseableSample - starting samplest.lifecycle.AutoCloseableSample@17fdbd39
<< RESPONSE
200
2020-04-28 21:15:38,306 [qtp1225970981-189] [admin     ] INFO  restx.StdRestxMainRouter - << [RESTX REQUEST] GET /uuids/random >> OK - 5.831 ms
2020-04-28 21:15:38,307 [qtp1225970981-189] [          ] INFO  s.lifecycle.AutoCloseableSample - closing samplest.lifecycle.AutoCloseableSample@17fdbd39
c81ed8c6-dc40-47c5-ae29-93a375f7b77c
checked GET /uuids/random -- 132.7 ms
dropping database mongodb://localhost:27017/testing-mongo-test-1588108538174-2
2020-04-28 21:15:38,309 [main            ] [          ] INFO  org.mongodb.driver.cluster - Cluster created with settings {hosts=[localhost:27017], mode=SINGLE, requiredClusterType=UNKNOWN, serverSelectionTimeout='30000 ms', maxWaitQueueSize=500}
2020-04-28 21:15:38,311 [cluster-ClusterId{value='5ea89cfa533eb82c6c478c8c', description='null'}-localhost:27017] [          ] INFO  org.mongodb.driver.cluster - Exception in monitor thread while connecting to server localhost:27017
com.mongodb.MongoSocketOpenException: Exception opening socket
	at com.mongodb.internal.connection.SocketStream.open(SocketStream.java:70)
	at com.mongodb.internal.connection.InternalStreamConnection.open(InternalStreamConnection.java:128)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.run(DefaultServerMonitor.java:117)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connexion refusée (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:607)
	at com.mongodb.internal.connection.SocketStreamHelper.initialize(SocketStreamHelper.java:64)
	at com.mongodb.internal.connection.SocketStream.initializeSocket(SocketStream.java:79)
	at com.mongodb.internal.connection.SocketStream.open(SocketStream.java:65)
	... 3 common frames omitted
2020-04-28 21:15:38,313 [main            ] [          ] INFO  org.mongodb.driver.cluster - Cluster description not yet available. Waiting for 30000 ms before timing out
Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 30.198 sec <<< FAILURE! - in samplest.core.UUIDResourceSpecTest
samplest.core.UUIDResourceSpecTest  Time elapsed: 30.143 sec  <<< ERROR!
com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting to connect. Client view of cluster state is {type=UNKNOWN, servers=[{address=localhost:27017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketOpenException: Exception opening socket}, caused by {java.net.ConnectException: Connexion refusée (Connection refused)}}]
	at com.mongodb.internal.connection.BaseCluster.getDescription(BaseCluster.java:182)
	at com.mongodb.internal.connection.SingleServerCluster.getDescription(SingleServerCluster.java:41)
	at com.mongodb.client.internal.MongoClientDelegate.getConnectedClusterDescription(MongoClientDelegate.java:152)
	at com.mongodb.client.internal.MongoClientDelegate.createClientSession(MongoClientDelegate.java:103)
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.getClientSession(MongoClientDelegate.java:284)
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:208)
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:182)
	at com.mongodb.DB.dropDatabase(DB.java:248)
	at com.mongodb.Mongo.dropDatabase(Mongo.java:550)
	at restx.jongo.specs.tests.GivenJongoCollectionSpecRuleSupplier$GivenJongoCollectionSpecRule.onTearDown(GivenJongoCollectionSpecRuleSupplier.java:66)
	at restx.tests.RestxSpecRunner.dispose(RestxSpecRunner.java:88)
	at restx.tests.RestxSpecRule.beforeServerStop(RestxSpecRule.java:82)
	at restx.tests.RestxServerRule.tearDown(RestxServerRule.java:60)
	at restx.tests.RestxServerRule$1.evaluate(RestxServerRule.java:77)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
```

# Call hierarchy

Right after server startup, we see the following log line:
```
using db mongodb://localhost:27017/testing-mongo-test-1588108538174-2
```
This log is generated by `restx.jongo.specs.tests.GivenJongoCollectionSpecRule.onSetup()`.

Doing a call hierarchy show this method is called at one place: `restx.tests.RestxSpecRunner.getSpecRulesForCurrentBlade()`:

```
private synchronized Iterable<GivenSpecRule> getSpecRulesForCurrentBlade() {
    String currentBlade = Blade.current();
    Iterable<GivenSpecRule> givenSpecRules = givenRulesPerBlade.get(currentBlade);
    if (givenSpecRules == null) {
        givenSpecRules = newArrayList(transform(givenSpecRuleSuppliers,
	        Suppliers.<GivenSpecRule>supplierFunction()));
        for (GivenSpecRule givenSpecRule : givenSpecRules) {
	    givenSpecRule.onSetup(contextLocal(bladeContextId(currentBlade)));
        }

        givenRulesPerBlade.put(currentBlade, givenSpecRules);
    }
    return givenSpecRules;
}
```

We get all the `givenSpecRuleSuppliers` we have in this object field and we call `onSetup()` for each.

This field is set in `RestxSpecRunner` constructor:
```
givenSpecRuleSuppliers = factory.queryByClass(GivenSpecRuleSupplier.class).findAsComponents();
```

`restx.factory.Factory.queryByClass(Class<T> componentClass)` instantiate a `ClassQuery<T>`. Then `findAsComponents()` calls the following inner methods `find()` > `doFind()`.

This method as the following implementation for `ClassQuery<T>`:
```
@Override
protected Set<NamedComponent<T>> doFind() {
    Set<NamedComponent<T>> components = Sets.newTreeSet(NAMED_COMPONENT_COMPARATOR);
    Factory factory = factory();
    for (Name<T> tName : factory.collectAllBuildableNames(componentClass)) {
        components.addAll(factory.queryByName(tName).optional().find());
    }

    return components;
}
```

`factory.collectAllBuildableNames(componentClass)` returns a `GivenJongoCollectionSpecRuleSupplier`:

```
private <T> Set<Name<T>> collectAllBuildableNames(Class<T> componentClass) {
    Set<Name<T>> buildableNames = Sets.newLinkedHashSet();
    for (FactoryMachine machine : machines) {
        buildableNames.addAll(nameBuildableComponents(machine, componentClass));
    }
    return buildableNames;
}
```

`machines` contains a `restx.jongo.specs.tests.GivenJongoCollectionSpecRuleSupplierFactoryMachine`. This field is passed to the `Factory` constructor. It looks like the constructor is just doing a sort on it. While debugging in the `collectAllBuildableNames()` method, I do see that `usedServiceLoader` field is `true`.

# Factory constructor

When setting a new breakpoint at `Factory.java:713` and relaunching this specific test. I see `GivenJongoCollectionSpecRuleSupplierFactoryMachine` in `Multimap<String, FactoryMachine> machines` under the `"ServiceLoader"` key.

I see the following call stack in Intellij Idea:
```
<init>:713, Factory (restx.factory)
<init>:37, Factory (restx.factory)
build:310, Factory$Builder (restx.factory)
getInstance:113, Factory (restx.factory)
<init>:27, RestxSpecRule (restx.tests)
<clinit>:14, UUIDResourceSpecTest (samplest.core)
```

Going up in the call hierarchy, at `build:310, Factory$Builder (restx.factory)`. I can see ServiceLoader machines being loaded with the following line:

```
machines.putAll(SERVICE_LOADER, FactoryMachinesServiceLoader.getMachines());
```

`Factory.FactoryMachinesServiceLoader.getMachines()` call `ServiceLoader.load(FactoryMachine.class)`. It find `GivenJongoCollectionSpecRuleSupplierFactoryMachine` because it is in the class path (we have it's module as a dependency).

The way all things are intricated, I honestly do not know how this could be solved. Having changes in `Factory` which allow to execute conditionnaly `FactoryMachine`s implementations ?

I'll gladly take any documentation on the `Factory` itself or any recommendation for fixing this. I haven't looked at the code long enough to see clearly what could be done respecting the initial principles of these classes.

**Comment by nhumblot** [comment]
> About the `MongoRestxSpecTestsRunner`, I still think we should find a better way. I think Mongo should be started only once for all tests - and the rule would just pick a up a new db for each test.
> 
> I think this implies moving to junit 5 which has @beforeall and @afterall and reviewing the logic (mongo embed rule should start and stop a db only if one is not already started globally by a beforeall). That's more work, but it's much cleaner.
> 
> WDYT?

I agree with you, I think it is safe to start only one Mongo instance, then execute all the tests. I've seen multiple times that tests relying on other tests is an anti-pattern, even for integration tests (IT). So if an IT mutates or creates data, this specific data should not be used in other ITs.

This behaviour will be better performance wise. 

Is moving to JUnit 5 going to break the API ? In case of, it looks like we can do that with JUnit 4 as well with [`RunListener`](https://junit.org/junit4/javadoc/latest/org/junit/runner/notification/RunListener.html) ([example](https://stackoverflow.com/a/14773170) - comments of this answer are interesting as well regarding process interruption). Has to be validated if it works in all configurations (with maven execution but also in IDE). I have not tested it.

If we can move to JUnit 5 without breaking anything (or if you are ok for it to break), it is not an issue for me, it is always a good thing to upgrade dependencies so I'm definitely ok with it.

File: restx-samplest/src/main/java/samplest/autostartable/AutoStartableTestComponent.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nhumblot) [reply restx-samplest/src/main/java/samplest/autostartable/AutoStartableTestComponent.java:19]: Hi,

I first tried to use a new `Factory` but I couldn't make it work. [Here is a change with a new Factory](https://github.com/nhumblot/restx/commit/1865bc87aca19e1c8765c8997042fb3b6ac76d51).

When launching a `mvn verify`, the test is failing.
```
Failed tests: 
  AutoStartableTest.should_handle_auto_startable_in_dev_mode:33 expected:<...called: 1 started: 1[ closed: 0 instanciated: ]1 serverId: Jetty8#1...> but was:<...called: 1 started: 1[11 closed: 110 instanciated: 11]1 serverId: Jetty8#1...>
Tests run: 83, Failures: 1, Errors: 0, Skipped: 0
```
Test log:
```
should_handle_auto_startable_in_dev_mode(samplest.autostartable.AutoStartableTest)  Time elapsed: 6.885 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<...called: 1 started: 1[ closed: 0 instanciated: ]1 serverId: Jetty8#1...> but was:<...called: 1 started: 1[11 closed: 110 instanciated: 11]1 serverId: Jetty8#1...>
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at samplest.autostartable.AutoStartableTest.should_handle_auto_startable_in_dev_mode(AutoStartableTest.java:33)
```


The test is not getting the expected count of `started`, `closed` and `instantiated` values.

One thing to note, when this test is executed alone, it passes.

I think it fails in `mvn verify` because we execute multiple tests. In this situation, `AutoStartableTest` is not the first test to run. So we start a server multiple times but JUnit keep the same class loader so static values are kept between tests.

So it means if we need to keep these fields static, we should also execute each unit test with a separate class loader ? This way, static fields will be reset and we can enforce the value ?

Or another solution, seems simpler and better to me, would be, at the start of the test, we get the `started`, `closed` and `instantiated` values, store them as local variables during the test and we use them as base values instead of supposing these values are all at 0 before we start the test.

After your message, I understand your concern, and I think I see the issue you are pointing. Is it because the dev mode throw and create components after each request ?

Please tell me if one of the two solutions exposed satisfy you and which one has your preference. I can work on that and update this pull request accordingly.

[the remark this answers is not in the document]
Comment (xhanin) [reply restx-samplest/src/main/java/samplest/autostartable/AutoStartableTestComponent.java:19]: You're right, thanks for this investigation.

I think the simplest solution would be to store the values before the test and compare with them, or add a `reset` method on AutoStartable which sets everything to 0, and call it in the test. Neither of these solutions would make the test able to run in parallel but at least it should solve the problem and still perform the intent of the test.

**Comment by xhanin** [comment]
I think moving to junit5 in the samplest would be fine, it's only a sample so it won't affect people using restx. But using a runlistener seems to be simpler to implement, so I think would go fo it.

And indeed moving away from EmbedMongo would be a good thing IMO, testcontainers is really cool. But it doesn't change the fact that it's better to start a single Mongo server, therefore I think ATM I would stick with EmbedMongo just to go on "quickly" with the test fix.

File: restx-samplest/src/main/java/samplest/autostartable/AutoStartableTestComponent.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nhumblot) [reply restx-samplest/src/main/java/samplest/autostartable/AutoStartableTestComponent.java]: This test has been fixed accordingly to our discussion. Change in 69ec801f0dc3dede6a324a95f62c13c4dc8a077f. You can add comments on this change if needed.

I updated to the current version of `master`, rebased this branch on it, and dropped the previous commit on `AutoStartableTest` fix that generated this comment.

I will start working on MongoDB startup onward.

[the remark this answers is not in the document]
Comment (xhanin) [reply restx-samplest/src/main/java/samplest/autostartable/AutoStartableTestComponent.java]: Perfect!

**Comment by nhumblot** [comment]
Hi @xhanin, I provided some changes:

I created a `MongoRestxSpecTestsListener` for launching a Mongo database before tests execution and shutting it down when all tests have been executed. This listener has been added in the `maven-surefire-plugin` configuration of `restx-samplest`. This listener is the only public exposure.

To prevent conflict between this new listener and the already existing `MongoRestxSpecTestsRunner`, they both register to a `EmbedMongoClientPool`.

This pool references all instances needing a specific database version. Database is launched after first check in and closed after the last check out from the pool.

This way, the Listener and the Runner goes well together. A `MongoRestxSpecTestsRunner` won't start a new database if a `MongoRestxSpecTestsListener` did it previously for the same version. But the Runner will still start the database if there is no listener. Also, a Runner won't cause the shut down of a mongod process if other runners or a listener registered to it.

`EmbedMongoFactory` guarantees two instances requesting the same Mongo version will get the same mongoUri and will therefore register to the same client pool.

Because of this change, it is no longer needed to instantiate a new factory in `RestxSpecTestsRunner` so I dropped the commit which raised a comment on your side during the first review.

Performing a `mvn verify` with jdk8 on `restx-parent` ends up with success.

Please give your feedback regarding my implementation.

One thing to note, `MongoRestxSpecTestsListener` does not allow to pick a version, it gets the production one, this could be an issue in some cases.

I will wait for your feedback so we can finalize this before working on a documentation update.

File: restx-jongo-specs-tests/src/main/java/restx/jongo/specs/tests/EmbedMongoFactory.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            keys.put(version, key);
+        }
+    }
+
+    PoolKey getPoolKey(Version.Main version) throws IOException {
+        if (Objects.isNull(keys.get(version))) {
+            newPoolKey(version);
+        }
+        return keys.get(version);
+    }
```
Comment (xhanin) [inline restx-jongo-specs-tests/src/main/java/restx/jongo/specs/tests/EmbedMongoFactory.java:41]: We have a race condition here: 2 threads can check the isNull in parallel, and then call newPoolKey in parallel. I think you should either synchronize the block or use ConcurrentMap putIfAbsent
